File: ch12/ring.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

type Node struct {
	ID   int
	Chan chan string
	Succ *Node
}

func main() {
	n, _ := strconv.Atoi(os.Args[1])
	m, _ := strconv.Atoi(os.Args[2])
	debug := os.Args[3] == "true"
	benchmark(n, m, debug)
}

func benchmark(n int, m int, debug bool) {
	// create n nodes
	nodes := make([]*Node, 0)
	for i := 0; i < n; i++ {
		nodes = append(nodes, &Node{i, make(chan string), nil})
	}

	// connect the nodes
	for i := 0; i < n; i++ {
		node := nodes[i]
		if i < n-1 {
			node.Succ = nodes[i+1]
		}
	}

	// run the nodes in separate Goroutines
	for i := 0; i < n-1; i++ {
		go forward(nodes[i], debug)
	}

	// initiate the ring call
	start := time.Now()
	go func(ch chan<- string) {
		for j := 0; j < m; j++ {
			ch <- fmt.Sprintf("hello %d", j)
		}
	}(nodes[0].Chan)

	// drain all the channels and measure time
	for j := 0; j < m; j++ {
		message := <-nodes[n-1].Chan
		if debug {
			fmt.Printf("'%s' forwarded from %d to %d\n", message, n-2, n-1)
		}
	}
	fmt.Printf("forwarded %d messages %d times in %v\n", m, n, time.Since(start))
}

func forward(node *Node, debug bool) {
	for {
		message := <-node.Chan
		if debug {
			fmt.Printf("'%s' forwarded from %d to %d\n", message, node.ID-1, node.ID)
		}
		node.Succ.Chan <- message
	}
}

File: ch12/ring.erl
-module(ring).
-export([benchmark/3, forward/3]).

benchmark(N, M, Debug) ->
    % create and run n+1 nodes (one pseudo)
    NodeIDs = lists:seq(0, N),
    Successors = lists:zip(NodeIDs, tl(NodeIDs), {pad, {N + 1, 0}}),
    PidPairs = lists:foldr(fun(Pair, Acc) -> connect(Pair, Acc, Debug) end, [], Successors),

    % initiate the ring call
    {_ , First} = hd(PidPairs),
    statistics(wall_clock),
    lists:foreach(fun(I) -> ring_call(First, I) end, lists:seq(1, M)),

    % drain the inbox and measure time
    receive
        _ ->
            {_, Time} = statistics(wall_clock),
            io:format("forwarded ~b messages ~b times in ~fs~n", [M, N, Time / 1000])
    end.

ring_call(Pid, I) ->
    Pid ! io_lib:format("hello ~b", [I]).

connect({L, 0}, Acc, _) ->
    [{L, self()} | Acc];
connect({L, _}, Acc, Debug) ->
    [{_, Succ} | _] = Acc,
    [{L, spawn(ring, forward, [L, Succ, Debug])} | Acc].

forward(ID, Succ, Debug) ->
    receive
        Message ->
            if 
                Debug -> io:format("'~s' forwarded from ~b to ~b~n", [Message, ID-1, ID]);
                true -> ok
            end,
            Succ ! Message
    end,
    forward(ID, Succ, Debug).

